File: core/src/main/java/core2/vector/extensions/UuidType.java
package core2.vector.extensions;

import org.apache.arrow.memory.BufferAllocator;
import org.apache.arrow.vector.FieldVector;
import org.apache.arrow.vector.types.pojo.ArrowType;
import org.apache.arrow.vector.types.pojo.ExtensionTypeRegistry;
import org.apache.arrow.vector.types.pojo.FieldType;

public class UuidType extends XtExtensionType {
    public static final UuidType INSTANCE = new UuidType();

    static {
        ExtensionTypeRegistry.register(INSTANCE);
    }

    public UuidType() {
        super("uuid", new FixedSizeBinary(16));
    }

    @Override
    protected ArrowType deserialize(String serializedData) {
        return INSTANCE;
    }

    @Override
    public FieldVector getNewVector(String name, FieldType fieldType, BufferAllocator allocator) {
        return new UuidVector(name, allocator, fieldType);
    }
}

File: api/src/main/kotlin/xtdb/vector/extensions/UuidType.kt
package xtdb.vector.extensions

import org.apache.arrow.memory.BufferAllocator
import org.apache.arrow.vector.FieldVector
import org.apache.arrow.vector.types.pojo.ExtensionTypeRegistry
import org.apache.arrow.vector.types.pojo.FieldType

object UuidType : XtExtensionType("uuid", FixedSizeBinary(16)) {
    init {
        ExtensionTypeRegistry.register(this)
    }

    override fun deserialize(serializedData: String) = this

    override fun getNewVector(name: String, fieldType: FieldType, allocator: BufferAllocator) =
        UuidVector(name, allocator, fieldType)
}
